{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      " * Serving Flask app '__main__' (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [26/Oct/2022 14:50:38] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [26/Oct/2022 14:50:38] \"GET /_dash-layout HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [26/Oct/2022 14:50:38] \"GET /_dash-dependencies HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [26/Oct/2022 14:50:38] \"GET /_favicon.ico?v=1.19.0 HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 2073, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1519, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1517, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\flask\\app.py\", line 1503, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1078, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\dash\\dash.py\", line 1009, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\ljens\\AppData\\Local\\Temp\\ipykernel_17084\\1559283082.py\", line 56, in get_pie_chart\n",
      "    fig = px.pie(spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 1466, in pie\n",
      "    layout_patch=layout_patch,\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1406, in build_dataframe\n",
      "    args, wide_mode, var_name, value_name\n",
      "  File \"c:\\Users\\ljens\\anaconda3\\envs\\epi_methods\\lib\\site-packages\\plotly\\express\\_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'names' is not the name of a column in 'data_frame'. Expected one of ['Unnamed: 0', 'Flight Number', 'Launch Site', 'class', 'Payload Mass (kg)', 'Booster Version', 'Booster Version Category'] but received: pie chart name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [26/Oct/2022 14:50:38] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    }
   ],
   "source": [
    "# Import required libraries\n",
    "import pandas as pd\n",
    "import dash\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "from dash.dependencies import Input, Output\n",
    "import plotly.express as px\n",
    "\n",
    "# Read the airline data into pandas dataframe\n",
    "spacex_df = pd.read_csv(\"C:\\\\Users\\\\ljens\\\\projects\\\\coursera\\\\capstone\\\\spacex_launch_dash.csv\")\n",
    "max_payload = spacex_df['Payload Mass (kg)'].max()\n",
    "min_payload = spacex_df['Payload Mass (kg)'].min()\n",
    "\n",
    "# Create a dash application\n",
    "app = dash.Dash(__name__)\n",
    "\n",
    "# Create an app layout\n",
    "app.layout = html.Div(children=[html.H1('SpaceX Launch Records Dashboard',\n",
    "                                        style={'textAlign': 'center', 'color': '#503D36',\n",
    "                                               'font-size': 40}),\n",
    "                                # TASK 1: Add a dropdown list to enable Launch Site selection\n",
    "                                # The default select value is for ALL sites\n",
    "                                 dcc.Dropdown(id='site-dropdown',options=[{'label': 'All Sites', 'value': 'ALL'},{'label':'CCAFS LC-40','value':'CCAFS LC-40'},{'label':'VAFB SLC-4E','value':'VAFB SLC-4E'},{'label':'KSC LC-39A','value':'KSC LC-39A'},{'label':'CCAFS SLC-40','value':'CCAFS SLC-40'}], placeholder = 'Select a Launch Site here', searchable=True),\n",
    "                                html.Br(),\n",
    "\n",
    "                                # TASK 2: Add a pie chart to show the total successful launches count for all sites\n",
    "                                # If a specific launch site was selected, show the Success vs. Failed counts for the site\n",
    "\n",
    "                                # Function decorator to specify function input and output\n",
    "                                @app.callback(Output(component_id='success-pie-chart', component_property='figure'),\n",
    "                                            Input(component_id='site-dropdown', component_property='value'))\n",
    "                                def get_pie_chart(entered_site):\n",
    "                                    filtered_df = spacex_df\n",
    "                                    if entered_site == 'ALL':\n",
    "                                        fig = px.pie(data = spacex_df, values='class', \n",
    "                                        names='pie chart names', \n",
    "                                        title='title')\n",
    "                                        return fig\n",
    "                                    else:\n",
    "                                        # return the outcomes piechart for a selected site\n",
    "                                        fig = px.pie(data = spacex_df.loc[spacex_df['Launch Site']==entered_site, :], values='class',names='pie chart name', title='title')\n",
    "                                        return fig  \n",
    "\n",
    "                                                                html.Div(dcc.Graph(id='success-pie-chart')),\n",
    "                                                                html.Br(),\n",
    "\n",
    "                                                                html.P(\"Payload range (Kg):\"),\n",
    "                                                                # TASK 3: Add a slider to select payload range\n",
    "                                                                #dcc.RangeSlider(id='payload-slider',...)\n",
    "\n",
    "                                                                # TASK 4: Add a scatter chart to show the correlation between payload and launch success\n",
    "                                                                html.Div(dcc.Graph(id='success-payload-scatter-chart')),\n",
    "                                                                ])\n",
    "\n",
    "# TASK 2:\n",
    "# Add a callback function for `site-dropdown` as input, `success-pie-chart` as output\n",
    "\n",
    "# TASK 4:\n",
    "# Add a callback function for `site-dropdown` and `payload-slider` as inputs, `success-payload-scatter-chart` as output\n",
    "\n",
    "\n",
    "# Run the app\n",
    "if __name__ == '__main__':\n",
    "    app.run_server()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CCAFS LC-40', 'VAFB SLC-4E', 'KSC LC-39A', 'CCAFS SLC-40'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spacex_df['Launch Site'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.11 ('epi_methods')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "32d945d47d07914d4d67bd1543a3e0968c007539a3cf4473a9b272e21f3ba6ca"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
